src/db.rs

```diff
@@ -660,41 +711,124 @@ impl DB {
         Ok(convert_values(values, values_sizes))
     }
 
-    pub fn create_cf<N: AsRef<str>>(&mut self, name: N, opts: &Options) -> Result<(), Error> {
-        let cf_name = if let Ok(c) = CString::new(name.as_ref().as_bytes()) {
+    fn create_inner_cf_handle(
+        &self,
+        name: &str,
+        opts: &Options,
+    ) -> Result<*mut ffi::rocksdb_column_family_handle_t, Error> {
+        let cf_name = if let Ok(c) = CString::new(name.as_bytes()) {
             c
         } else {
             return Err(Error::new(
                 "Failed to convert path to CString when creating cf".to_owned(),
             ));
         };
-        unsafe {
-            let inner = ffi_try!(ffi::rocksdb_create_column_family(
+        Ok(unsafe {
+            ffi_try!(ffi::rocksdb_create_column_family(
                 self.inner,
                 opts.inner,
                 cf_name.as_ptr(),
-            ));
+            ))
+        })
+    }
 
-            self.cfs
-                .insert(name.as_ref().to_string(), ColumnFamily { inner });
+    /// Creates column family with given name by internally locking the inner column
+    /// family map. This avoids needing `&mut self` reference, but is only safe to
+    /// use if the database was opened with multi-threaded config
+    pub fn create_cf_multi_threaded<N: AsRef<str>>(
+        &self,
+        name: N,
+        opts: &Options,
+    ) -> Result<(), Error> {
+        let inner = self.create_inner_cf_handle(name.as_ref(), opts)?;
+        match &self.cfs {
+            ColumnFamilyHandles::MultiThread(cf_map) => cf_map
+                .write()
+                .unwrap()
+                .insert(name.as_ref().to_string(), Arc::new(ColumnFamily { inner })),
+            ColumnFamilyHandles::SingleThread(_) => {
+                panic!("Not available on SingleThread implementation");
+            }
         };
         Ok(())
     }
 
+    /// Creates column family with given name and options
+    pub fn create_cf<N: AsRef<str>>(&mut self, name: N, opts: &Options) -> Result<(), Error> {
+        let inner = self.create_inner_cf_handle(name.as_ref(), opts)?;
+        match &mut self.cfs {
+            ColumnFamilyHandles::MultiThread(_) => {
+                panic!("Not available on MultiThread implementation")
```

Comment:
Sorry for my nicety, but I have a feeling that it's not good to panic here. What about to bound an invoking this with trait ??? I'll try to provide the idea with code soon.

Reply:
@aleksuss Hi, again. I want this to move forward :) (cc: @carllin)

> What about to bound an invoke of this with trait ??? I'll try to provide the idea with code soon.

Maybe, were you thinking about like this?: https://github.com/ryoqun/rust-rocksdb/commit/1be386d3d388901398b29e67ecc5f56d6f59f9cf

Sorry for very rough code. But this eliminates bunch of various concerns arisen from this pr's review:

- ColumnFamily is bounded in the case of MultiThread with zero runtime overhead (I even stopped using Arc)
- will have no `panic!`s (if everything is correctly rewritten)
- keeps api compatibility and no multithread methods.
- no enum runtime variant branching (although this is cheap)
- downside: a bit of internal code churn (I'll promise I'll work hard to fix them all)

If this direction is really to go from maintainer's perspective, I'll finish this very quickly. :)

Reply:
> Maybe, were you thinking about like this?: ryoqun@1be386d
Yes. That is exactly what I meant 👍🏻
Unfortunately, I have no time now to provide my vision on it. But your approach is very very near to my vision. Let's try to implement it 🙂

Reply:
@aleksuss as promised, I created the cleaned-up pr: https://github.com/rust-rocksdb/rust-rocksdb/pull/506 Please have a look at your convenient time. :)